### ISSUE TITLE
private event handlers, change watchers, databinding doesn't work

### ISSUE THREAD

**Issue by DartBot** [dart-lang/smoke#11]
<a href="https://github.com/kevmoo"><img src="https://avatars.githubusercontent.com/u/17034?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [kevmoo](https://github.com/kevmoo)**
_Originally opened as dart-lang/sdk#13881_

---

Uncaught Illegal argument(s): &quot;_something&quot; is a private identifier

It seems like a name is inferred from private members ending with 'Changed'

This is a regression since 0.7.6

**Comment by DartBot** [comment]
<a href="https://github.com/kevmoo"><img src="https://avatars.githubusercontent.com/u/17034?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [kevmoo](https://github.com/kevmoo)**

---

Here's a commit demonstrating my work-around.

https://github.com/GoogleChrome/notational_velocity.dart/commit/8312bdcd3ea6baca88551379f3a4b70334c97f45

Renaming the private methods to end in 'Mutated' instead of 'Changed' eliminates the runtime error

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

Right, "Changed" is a Polymer feature for setting up an observer.
http://www.polymer-project.org/polymer.html#change-watchers

... we can't do the right thing here until Dart2js supports private symbols :(. Observable has same issue.

---

_Added [Library-Polymer](../labels/Library-Polymer), [Triaged](../labels/Triaged) labels._
_Marked this as being blocked by dart-lang/sdk#13355._

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

---

_Added this to the [M8](../milestones/M8) milestone._

**Comment by DartBot** [comment]
<a href="https://github.com/kevmoo"><img src="https://avatars.githubusercontent.com/u/17034?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [kevmoo](https://github.com/kevmoo)**

---

It feels *very* weird that the names of private methods in my classes would cause runtime errors.

Is there a check somewhere looking for methods that end in 'Changed'? Could this check add the condition that the method not start with "_" -- and is therefor private?

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

---

_Removed this from the [M8](../milestones/M8) milestone._
_Added this to the [Later](../milestones/Later) milestone._
_Removed [Priority-Unassigned](../labels/Priority-Unassigned) label._
_Added [Priority-Medium](../labels/Priority-Medium) label._

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

Issue dart-lang/sdk#14191 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

Issue dart-lang/sdk#13559 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

would be nice to fix in M8 but blocked on dart2js not-implemented-feature.

---

_Marked this as being blocked by dart-lang/sdk#14231._
_Changed the title to: "private event handlers, change watchers, databinding don't work"._

**Comment by DartBot** [comment]
<a href="https://github.com/ussuri"><img src="https://avatars.githubusercontent.com/u/5606182?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [ussuri](https://github.com/ussuri)**

---

Re merging Issue dart-lang/sdk#14191 into this.

14191 indeed seems to have the same root cause (lack of support for private members), but the practical consequences are different: this one at least produces a runtime error, allowing the user to track it down, while with 14191 the unsupported usage is silently ignored. So while this can wait for the root cause to be fixed, it would be very desirable to at least add some diagnostic to the 14191 use case in the meantime.

Also, 14191 is not about dart2js at all: it doesn't work in Dartium either.

**Comment by DartBot** [comment]
<a href="https://github.com/ussuri"><img src="https://avatars.githubusercontent.com/u/5606182?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [ussuri](https://github.com/ussuri)**

---

Ok, with the changed title this looks better :)

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

> Also, 14191 is not about dart2js at all: it doesn't work in Dartium either.

Right, it doesn't work on either impl because we're waiting on a new API MirrorSystem.getSymbol. They just implemented it in Dart VM but not in dart2js yet. I can't use the new API until it works in both places. Hope that clarifies.

**Comment by DartBot** [comment]
<a href="https://github.com/kevmoo"><img src="https://avatars.githubusercontent.com/u/17034?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [kevmoo](https://github.com/kevmoo)**

---

_Changed the title to: "private event handlers, change watchers, databinding doesn't work"._

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

---

_Removed this from the [Later](../milestones/Later) milestone._
_Added this to the [M9](../milestones/M9) milestone._

**Comment by DartBot** [comment]
<a href="https://github.com/clayberg"><img src="https://avatars.githubusercontent.com/u/8616782?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [clayberg](https://github.com/clayberg)**

---

_Removed this from the [M9](../milestones/M9) milestone._
_Added this to the [1.1](../milestones/1.1) milestone._

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

---

_Removed this from the [1.1](../milestones/1.1) milestone._
_Added this to the [1.2](../milestones/1.2) milestone._

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

---

Siggi, I wonder if we could fix this in Smoke? (or it's already fixed?)

---

_Removed [Area-Polymer](../labels/Area-Polymer), [Library-Polymer](../labels/Library-Polymer) labels._
_Added [Area-Pkg](../labels/Area-Pkg), [Pkg-Smoke](../labels/Pkg-Smoke) labels._

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

---

it's possible, but not trivial because it requires changing how we generate code into the original source files (the way we generate getters today assumes that everything is visible from outside the library).

I happen to have a bug tracking this enhancement though. Marking as duplicate.

---

_Added [Duplicate](../labels/Duplicate) label._
_Marked as being merged into dart-lang/polymer-dart#390._
